### PR TITLE
Run CI in parallel

### PR DESCRIPTION
**Pull request by Vlatombe**
Was able to go down to an overall 27m (from initial 44m)

Run CI in parallel to speed up builds.

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -3,34 +3,55 @@ properties([
...
-                        junit 'surefire-reports/*.xml'
-                    }
-                } finally {
-                    dir (WSTMP) {
-                        if (fileExists('kindlogs/docker-info.txt')) {
-                            archiveArtifacts 'kindlogs/'
+
+def splits
+stage('Determine splits') {
+    node('linux') {
```
Comment (jglick) [inline Jenkinsfile:9]: May want to use a container agent.

File: Jenkinsfile [hunk 2]
```diff
@@ -3,34 +3,55 @@ properties([
...
+        def split = splits[num]
+        def index = num + 1
+        branches["kind-${index}"] = {
+            node('docker') {
+                timeout(90) {
+                    checkout scm
+                    try {
+                        writeFile file: (split.includes ? "$WORKSPACE_TMP/includes.txt" : "$WORKSPACE_TMP/excludes.txt"), text: split.list.join("\n")
+                        writeFile file: (split.includes ? "$WORKSPACE_TMP/excludes.txt" : "$WORKSPACE_TMP/includes.txt"), text: ''
+                        sh './kind.sh -Dsurefire.includesFile="$WORKSPACE_TMP/includes.txt" -Dsurefire.excludesFile="$WORKSPACE_TMP/excludes.txt"'
```
Comment (jglick) [inline Jenkinsfile:29]: Hmm, so this winds up installing Kind twice per build. Expensive, but I guess worth the reduction in clock time.

File: test-in-k8s.sh [hunk 3]
```diff
@@ -9,21 +9,16 @@ kubectl get ns kubernetes-plugin-test || kubectl create ns kubernetes-plugin-tes
...
 port_offset=$RANDOM
-http_port=$((2000 + $port_offset))
-tcp_port=$((2001 + $port_offset))
+http_port=$((2000 + port_offset))
+tcp_port=$((2001 + port_offset))
 kubectl delete --ignore-not-found --now pod jenkins
 sed "s/@HTTP_PORT@/$http_port/g; s/@TCP_PORT@/$tcp_port/g" < test-in-k8s.yaml | kubectl apply -f -
 kubectl wait --for=condition=Ready --timeout=15m pod/jenkins
+# Copy temporary split files
+tar cf - "$WORKSPACE_TMP" | kubectl exec -i jenkins -- tar xf -
```
Comment (jglick) [inline test-in-k8s.sh:18]: Confused about how this works, without a `-C` arg or anything. Are we sharing absolute path names between the agent VM and the pod?

**Review by jglick — COMMENTED** [review]
Neat!

File: Jenkinsfile (hunk 1 above)
Comment (Vlatombe) [reply Jenkinsfile:9]: Yeah I wasn't sure of the right syntax in the context of `ci.jenkins.io`.

File: Jenkinsfile (hunk 2 above)
Comment (Vlatombe) [reply Jenkinsfile:29]: yeah the whole `kind` setup is ~2 min I think.

File: Jenkinsfile (hunk 1 above)
Comment (jglick) [reply Jenkinsfile:9]: You can use e.g. `maven-11`.

File: Jenkinsfile (hunk 2 above)
Comment (jglick) [reply Jenkinsfile:29]: Including downloading images? Not bad.

File: test-in-k8s.sh (hunk 3 above)
Comment (Vlatombe) [reply test-in-k8s.sh:18]: we're root inside `jenkins` container so we're effectively keeping the same absolute path. This avoids an extra path translation step when passing include/exclude options.

Comment (jglick) [reply test-in-k8s.sh:18]: And `tar x` just creates the missing directories like `/home/jenkins/workspace/es-plugin-PR_123@tmp`? Weird but OK if it works.

Comment (Vlatombe) [reply test-in-k8s.sh:18]: tar cf creates a tar with full path references, so we unpack these under `/` in the containe.r

File: Jenkinsfile (hunk 2 above)
Comment (lemeurherve) [reply Jenkinsfile:29]: We could include `kind` in the builder base image. WDYT?

Comment (jglick) [reply Jenkinsfile:29]: Well the thing is that the `kind` executable is just a few Mb; most of the download bandwidth is for the “Kindest node” images when creating a cluster.

Anyway we pin a specific version in plugin sources https://github.com/jenkinsci/kubernetes-plugin/blob/58918882cfb7096388049f60b4e269a8bfd5e10f/kind.sh#L7 and do really need to control exactly when that is updated (as part of a PR which may be adding or removing support for some K8s versions), so a binary baked into the VM would be useless.

Comment (lemeurherve) [reply Jenkinsfile:29]: Thanks for the detailed response @jglick

**Pull request merged**